python/preprocessor.py

```diff
@@ -0,0 +1,349 @@
+
+        self.weight_preprocessing_functions = [weight_preprocessing_function_map[name] for name in self.config[WEIGHT]]
+
+        logger.debug("Initializing Preprocessor: Done")
+
+    # preprocessor functions defined here
+    # they should be of the form
+    # arguments: feature_array, mask, observables, other keyword arguments
+    # returns: preprocessed array, modified observable list (in this order)
+
+    def _angle_to_sin_cos(self, feature_array, mask, observables, **args):
+        """
+        maps an angle to its sine and cosine
+
+        arguments
+        ---------
+        feature_array: 1d numpy array representing of some angle observable
+        mask: list of boolean, indicating which observables to be modified
+        observables: list of str, observable names indicating their position in the feature array
+
+        returns
+        -------
+        a 2d numpy array of the shape (number of events, number of observabless) with the original observables replaced by sine and cosine under the same name
+        """
+
+        constant = feature_array[:, ~mask] # part of feature array that will not get modified
+        modifying = feature_array[:, mask] # angles that will be encoded
+
+        feature_array = np.concatenate((constant, np.sin(modifying), np.cos(modifying)), axis = 1)
+
+        observables = np.concatenate((observables[~mask], observables[mask], observables[mask]))
+
+        return feature_array, observables
+
+    def _angle_to_sin(self, feature_array, mask, observables, **args):
+        """
+        maps an angle to sine of that angle
+
+        arguments
+        ---------
+        feature_array: 1d numpy array representing of some angle observable
+        mask: list of boolean, indicating which observables to be modified
+        observables: list of str, observable names indicating their position in the feature array
+
+        returns
+        -------
+        a 2d numpy array of the shape (number of events, number of observabless) with the original observables replaced by sine under the same name
+        """
+
+        feature_array[:, mask] = np.sin(feature_array[:, mask])
+
+        return feature_array, observables
+
+    def _angle_to_cos(self, feature_array, mask, observables, **args):
+        """
+        maps an angle to cosine of that angle
+
+        arguments
+        ---------
+        feature_array: 1d numpy array representing of some angle observable
+        mask: list of boolean, indicating which observables to be modified
+        observables: list of str, observable names indicating their position in the feature array
+
+        returns
+        -------
+        a 2d numpy array of the shape (number of events, number of observabless) with the original observables replaced by cosine under the same name
+        """
+
+        feature_array[:, mask] = np.cos(feature_array[:, mask])
+
+        return feature_array, observables
+
+    def _mask(self, observables, modify):
```

Comment:
Is `modify` initialized or assigned anywhere?

Reply:
It is written directly in preprocessor config. For example, if I would like to apply cosine map to "th_phi" and "tl_phi", the corresponding config is `"angle_to_sin_cos": ["th_phi", "tl_phi"]`, where the array part corresponds to `modify`.

Reply:
Is `modify` a global variable set somewhere?

Reply:
When the function `_mask` is called below, there are only two arguments right?

Reply:
Yes, only two arguments, the first being `observables` and the second named `list_of_observables` in `preprocess` correspond to `modify`.

I'm not entirely sure what the issue is here. I'm guessing you are saying that there are `self, observables, modify` in `_mask`'s arguments while `preprocess` calls `self._mask(observables, list_of_observable)`? Unlike c++, the first arguments is handled implicitly in python, so we just need to give it two arguments and ignore the `self`

Reply:
Oops sorry my bad... Should have read more carefully.